pkg/bpf/collection.go

```diff
@@ -41,6 +48,135 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
 	return spec, nil
 }
 
+func RemoveUnreachableTailcalls(spec *ebpf.CollectionSpec) error {
+	return removeUnreachableTailcalls(spec)
+}
+
+func removeUnreachableTailcalls(spec *ebpf.CollectionSpec) error {
+	type TailCall struct {
+		referenced bool
+		visited    bool
+		spec       *ebpf.ProgramSpec
+	}
+
+	entrypoints := make([]*ebpf.ProgramSpec, 0)
+	tailcalls := make(map[uint32]*TailCall)
+
+	for _, prog := range spec.Programs {
+		var id, slot uint32
+		// Consider any program that doesn't follow the tailcall naming convention
+		// x/y to be an entrypoint.
+		if _, err := fmt.Sscanf(prog.SectionName, "%d/%v", &id, &slot); err != nil {
+			entrypoints = append(entrypoints, prog)
+			continue
+		}
+
+		const (
+			cilium_calls_map = 2
+		)
+
+		// Any program that does follow the x/y naming convention but not part
+		// of the cilium_calls map is also considered an entrypoint.
+		if id != cilium_calls_map {
+			entrypoints = append(entrypoints, prog)
+			continue
+		}
+
+		tailcalls[slot] = &TailCall{
```

Comment:
Paranoid mode: check for duplicate slots? I guess that isn't possible since section names can't be the same?

Reply:
Hmm, we should not have duplicates since section names are unique, but with the scanf that accepts multiple encodings it can technically happen (`2/1` and `2/0x1`). Added extra guard rails can't hurt.